src/virtio/block.c

```diff
@@ -0,0 +1,463 @@
+#else
+#define LOG_BLOCK(...) do{}while(0)
+#endif
+
+#define LOG_BLOCK_ERR(...) do{ printf("VIRTIO(BLOCK)|ERROR: "); printf(__VA_ARGS__); }while(0)
+
+/* Maximum number of buffers in sddf data region */
+//@ericc: as long as this number is higher than any sddf_handler->data_size we should be fine
+#define SDDF_MAX_DATA_BUFFERS 4096
+
+/* Virtio blk configuration space */
+static struct virtio_blk_config virtio_blk_config;
+
+/* Data struct that handles allocation and freeing of fixed size data cells in sDDF memory region */
+static fsmem_t fsmem_data;
+static bitarray_t fsmem_avail_bitarr;
+static word_t fsmem_avail_bitarr_words[roundup_bits2words64(SDDF_MAX_DATA_BUFFERS)];
+
+/* Handle storing and retrieving of request data between virtIO and sDDF */
+typedef struct data {
+    uint16_t virtio_desc_head;
+    uintptr_t sddf_data;
+    uint16_t sddf_count;
+    uint32_t sddf_block_number;
+    uintptr_t virtio_data;
+    uint16_t virtio_data_size;
+    bool not_aligned;
+} data_t;
+static datastore_t datastore;
+static data_t datastore_data[SDDF_MAX_DATA_BUFFERS];
+static uint64_t datastore_nextfree[SDDF_MAX_DATA_BUFFERS];
+static bool datastore_used[SDDF_MAX_DATA_BUFFERS];
+
+static void virtio_blk_mmio_reset(struct virtio_device *dev)
+{
+    dev->vqs[VIRTIO_BLK_DEFAULT_VIRTQ].ready = 0;
+    dev->vqs[VIRTIO_BLK_DEFAULT_VIRTQ].last_idx = 0;
+}
+
+static int virtio_blk_mmio_get_device_features(struct virtio_device *dev, uint32_t *features)
+{
+    if (dev->data.Status & VIRTIO_CONFIG_S_FEATURES_OK) {
+        LOG_BLOCK_ERR("driver somehow wants to read device features after FEATURES_OK\n");
+    }
+
+    switch (dev->data.DeviceFeaturesSel) {
+        // feature bits 0 to 31
+        case 0:
+            *features = BIT_LOW(VIRTIO_BLK_F_FLUSH);
+            *features = *features | BIT_LOW(VIRTIO_BLK_F_BLK_SIZE);
+            break;
+        // features bits 32 to 63
+        case 1:
+            *features = BIT_HIGH(VIRTIO_F_VERSION_1);
+            break;
+        default:
+            LOG_BLOCK_ERR("driver sets DeviceFeaturesSel to 0x%x, which doesn't make sense\n", dev->data.DeviceFeaturesSel);
+            return 0;
+    }
+    return 1;
+}
+
+static int virtio_blk_mmio_set_driver_features(struct virtio_device *dev, uint32_t features)
+{
+    // According to virtio initialisation protocol,
+    // this should check what device features were set, and return the subset of features understood
+    // by the driver.
+    int success = 1;
+
+    uint32_t device_features = 0;
+    device_features = device_features | BIT_LOW(VIRTIO_BLK_F_FLUSH);
+    device_features = device_features | BIT_LOW(VIRTIO_BLK_F_BLK_SIZE);
+
+    switch (dev->data.DriverFeaturesSel) {
+        // feature bits 0 to 31
+        case 0:
+            success = (features == device_features);
+            break;
+        // features bits 32 to 63
+        case 1:
+            success = (features == BIT_HIGH(VIRTIO_F_VERSION_1));
```

Comment:
We follow the legacy virtIO?

Reply:
not offering the VIRTIO_F_VERSION_1 bit indicates it is legacy

Reply:
Ah sorry, I misunderstood this feature bit. Just looked at the spec and this looks correct.